python/fusion_engine_client/messages/measurement_details.py

```diff
@@ -0,0 +1,136 @@
+import struct
+
+from construct import Adapter, Struct, Int8ul, Padding
+import numpy as np
+
+from .timestamp import Timestamp, TimestampConstruct
+from ..utils.construct_utils import AutoEnum, ClassAdapter
+from ..utils.enum_utils import IntEnum
+
+
+class SensorDataSource(IntEnum):
+    """!
+    @brief The source of received sensor measurements, if known.
+    """
+    ## Data source not known.
+    UNKNOWN = 0
+    ## Sensor data captured internal to the device (embedded IMU, GNSS receiver, etc.).
+    INTERNAL = 1
+    ## Sensor data generated via hardware voltage signal (wheel tick, external event, etc.).
+    HARDWARE_IO = 2
+    ## Sensor data captured from a vehicle CAN bus.
+    CAN = 3
+    ## Sensor data provided over a serial connection.
+    SERIAL = 4
+    ## Sensor data provided over a network connection.
+    NETWORK = 5
+
+
+class SystemTimeSource(IntEnum):
+    """!
+    @brief The source of a @ref point_one::fusion_engine::messages::Timestamp used to represent the time of
+           applicability of an incoming sensor measurement.
+    """
+    ## Timestamp not valid.
+    INVALID = 0
+    ## Message timestamped in P1 time.
+    P1_TIME = 1
+    ## Message timestamped in system time, generated when received by the device.
+    TIMESTAMPED_ON_RECEPTION = 2
+    ## Message timestamp was generated from a monotonic clock of an external system.
+    SENDER_SYSTEM_TIME = 3
+    ## Message timestamped in GPS time, referenced to 1980/1/6.
+    GPS_TIME = 4
+
+
+class MeasurementDetails(object):
+    """!
+    @brief The time of applicability and additional information for an incoming sensor measurement.
+    """
+    Construct = Struct(
+        "measurement_time" / TimestampConstruct,
+        "measurement_time_source" / AutoEnum(Int8ul, SystemTimeSource),
+        "data_source" / AutoEnum(Int8ul, SensorDataSource),
+        Padding(2),
+        "p1_time" / TimestampConstruct,
+    )
+
+    _STRUCT = struct.Struct('<BB2x')
```

Comment:
I guess it doesn't hurt anything, but is there a specific purpose for having both the construct and struct implementation of the parsing?

Reply:
I see it's used inside the other messages that are specified with struct. I suppose you could still use the construct there, but I guess it's probably easier to just clean this up another day.

Reply:
Yeah my thoughts exactly.